pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -319,422 +320,469 @@ async def run(self):
+
+                    elif record.type in UPDATE_REPODATA:
+                        updateinfo_url = urljoin(remote_url, record.location_href)
+                        downloader = self.remote.get_downloader(url=updateinfo_url)
+                        downloaders.append([downloader.run()])
+
+                    elif record.type in COMPS_REPODATA:
+                        comps_url = urljoin(remote_url, record.location_href)
+                        comps_downloader = self.remote.get_downloader(url=comps_url)
+
+                    elif record.type in SKIP_REPODATA:
+                        continue
 
-                comps = libcomps.Comps()
-                comps.fromxml_f(comps_result.path)
+                    elif '_zck' in record.type:
+                        continue
 
-                with ProgressReport(message='Parsed Comps', code='parsing.comps') as comps_pb:
-                    comps_total = (
-                        len(comps.groups) + len(comps.categories) + len(comps.environments)
-                    )
-                    comps_pb.total = comps_total
-                    comps_pb.done = comps_total
-
-                if comps.langpacks:
-                    langpack_dict = PackageLangpacks.libcomps_to_dict(comps.langpacks)
-                    packagelangpack = PackageLangpacks(
-                        matches=strdict_to_dict(comps.langpacks),
-                        digest=dict_digest(langpack_dict)
-                    )
-                    dc = DeclarativeContent(content=packagelangpack)
-                    dc.extra_data = defaultdict(list)
-                    await self.put(dc)
+                    elif record.type in MODULAR_REPODATA:
+                        modules_url = urljoin(remote_url, record.location_href)
+                        modulemd_downloader = self.remote.get_downloader(url=modules_url)
+                        modulemd_results = await modulemd_downloader.run()
+
+                    elif record.type not in PACKAGE_DB_REPODATA:
+                        file_data = {record.checksum_type: record.checksum, "size": record.size}
+                        da = DeclarativeArtifact(
+                            artifact=Artifact(**file_data),
+                            url=urljoin(remote_url, record.location_href),
+                            relative_path=record.location_href,
+                            remote=self.remote,
+                            deferred_download=False
+                        )
+                        repo_metadata_file = RepoMetadataFile(
+                            data_type=record.type,
+                            checksum_type=record.checksum_type,
+                            checksum=record.checksum,
+                        )
+                        dc = DeclarativeContent(content=repo_metadata_file, d_artifacts=[da])
+                        await self.put(dc)
+
+                missing_type = set(PACKAGE_REPODATA) - main_types
+                if missing_type:
```

Comment:
if i understand correctly how mirrolist works is - is looks for the first available mirror and tries it. If there is some issue with the repodata it just fails. So once the download start from a url there is not way back.
There are some BZ opened around this behaviour and we agreed that amatej will inform us once they fix this( complicated change)
https://bugzilla.redhat.com/show_bug.cgi?id=1819188
https://bugzilla.redhat.com/show_bug.cgi?id=1819188
 I think we should follow same logic as currently mirrorlist has implemented. Thoughts?

Reply:
For the initial feature I think we should do the very simplest thing possible. I believe that would be:

1. Use the repodata from the very first available mirror that will serve it, and try to fetch all content from that mirror exclusively. If anything goes wrong in fetching content fail (just like Pulp normally would in cases where there is no mirrorlist).
2. Do not try to sort mirrors based on how fast or close they are
3. Do not try to fetch content from multiple mirrors

We can improve (2) and (3) later, but having even basic mirrorlist support is still a useful feature for users because we currently do not support it.

Reply:
After rpm meeting we decided that for now we would want to follow the current implementation of mirrorlist( aka fail if the the first fetched url fails during download)
In the future we could add this kind of enhancement where other urls would be tried if previous fail, but for that we'd rather implement a mirrorlist downloader to be more flexibile